<a href="https://colab.research.google.com/github/thfruchart/1nsi-2020/blob/master/Chap12/CORRECTION/EXERCICES_Trier_un_tableau.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercice 1

On souhaite compter le nombre d'opérations élémentaires réalisées lors du tri par sélection. 

Pour cela, on définit une fonction `nb_operation_tri_selection` qui :
* prend en argument un tableau t
* effectue le tri par sélection sur t
* renvoie le nombre d'opérations élémentaires effectuées : 
   * comparaison
   * affectation


1. Compléter le code ci-dessous pour remplir cet objectif.
2. Tester la fonction en calculant le nombre d'opérations nécessaires pour trier les tableaux suivants : 
   * [1,2,3]
   * [3,2,1]
   * [1,2,3,4,5,6,7,8,9,10,11,12]
3. Quel commentaire pouvez-vous faire? 

In [2]:
def nb_operation_tri_selection(t):
    cpt = 0
    for i in range(len(t)-1):
        i_min = i 
        cpt += 2 # affectation dans i et affectation dans i_min
        for k in range(i+1, len(t)):
            cpt += 2 # affectation dans k et comparaison
            if t[i_min] > t[k] :
                i_min = k
                cpt += 1 #affectation dans i_min
        t[i], t[i_min] = t[i_min], t[i] 
        cpt += 2 # 1 echange = 2 affectations 
    return cpt

In [3]:
nb_operation_tri_selection([1,2,3])

14

In [4]:
nb_operation_tri_selection([3,2,1])

16

In [5]:
nb_operation_tri_selection([1,2,3,4,5,6,7,8,9,10,11,12])

176

In [7]:
nb_operation_tri_selection([12,11,10,9,8,7,6,5,4,3,2,1])

212

#### 3)
Commentaire, pour un grand tableau, le nombre d'opération est **beaucoup plus** important que pour un petit tableau. 

[1,2,3,4,5,6,7,8,9,10,11,12] est 4 fois plus long que [1,2,3]

mais le nombre d'opérations trouvé : 176 > 4*14 = 56

Le tri par sélection nécessite un assez grand nombre d'opérations même lorsque le tableau est déjà trié !

# Exercice 2

On souhaite mesurer le temps de calcul pour exécuter le tri par sélection sur des tableaux aléatoires de différentes tailles

1. Ecrire une fonction `tableau_aleatoire(n)` qui
   * prend en argument un nombre entier `n >= 1`
   * renvoie un tableau de longueur n, contenant des flottants aléatoires entre 0 et 1
2. Evaluer le temps nécessaire pour trier :
   * un tableau de taille 100
   * un tableau de taille 200
   * un tableau de taille 500
3. Peut-on dire que le temps de calcul est environ proportionnel à la taille du tableau ? 

#### petits rappels sur les expressions aléatoires en python

In [8]:
from random import random

In [21]:
[random(), random(),random() ]

[0.08170457283431676, 0.37580199774006506, 0.13396858388371435]

In [22]:
from random import random

def tableau_aleatoire(n):
    t = [ 0.0 ] * n 
    for i in range(n):
        t[i] = random()
    return t 

# deuxième possibilité : tableau défini par compréhension
def tableau_aleatoire(n):
    return [random() for i in range(n)]

In [24]:
tableau_aleatoire(2)

[0.9614268618855472, 0.488533458451356]

#### tri d'un tableau aléatoire de taille 200

# Exercice 3

On considère un tableau d'entiers `tab` (type `list`) dont les éléments sont des **0** ou des **1**. 

On se propose de trier ce tableau selon l'algorithme suivant : 
* à chaque étape du tri, le tableau est constitué de trois zones consécutives, 
  * la première ne contenant que des 0, 
  * la seconde n'étant pas triée 
  * et la dernière ne contenant que des 1.
* on note `i` et `j` deux entiers tels que les indices de la zone non triée sont compris entre `i` et `j` inclus. 


```
[ Zone de 0 | Zone non triée | Zone de 1]
             i              j
```


Tant que la zone non triée n'est pas réduite à un seul élément, on regarde son premier élément:
* si cet élément vaut 0, on considère qu'il appartient désormais à la zone ne contenant que des 0 ;
*  si cet élément vaut 1, il est échangé avec le dernier élément de la zone non triée et on considère alors qu’il appartient à la zone ne contenant que des 1.


Dans tous les cas, la longueur de la zone non triée diminue de 1.

Compléter le code suivant pour réaliser cette fonction

In [ ]:
def tri(tab):
    #i est le premier indice de la zone non triee, j le dernier indice.
    #Au debut, la zone non triee est le tableau entier.
    i= ...
    j= ...
    while i != j :
        if tab[i]== 0:
            i= ...
        else :
            valeur = tab[j]
            tab[j] = ...
            ...
            j= ...
    ...
